### PR TITLE
RWA-2215: Add extra query parameters

### PR DESCRIPTION
**Pull request by rasikasri**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RWA-2215

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseService.java [hunk 1]
```diff
@@ -93,29 +92,28 @@ public Optional<String> findCaseId(String taskId) {
...
     }
 
-    public GetTasksResponse<Task> searchForTasks(SearchRequest searchRequest,
+    public GetTasksResponse<Task> searchForTasks(int firstResult,
+                                                 int maxResults,
+                                                 SearchRequest searchRequest,
                                                  AccessControlResponse accessControlResponse,
                                                  boolean granularPermissionResponseFeature) {
 
         List<RoleAssignment> roleAssignments = accessControlResponse.getRoleAssignments();
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseService.java:101]: I think it's a good idea to make sure we keep the original list of roleAssignments received from AM vs the filtered roleAssignments. This is for debugging or if in future it is needed in a different context.

Have to be a bit careful about copies in that it is done in a thread safe way.

**Review by raghera — APPROVED** [review]
Looked through the code and the Custom Repo tests.  All looks good for me ... just one comment which can be done separately if you like.

Worth getting Ben to look over your tests too which I am sure he does anyway.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseService.java (hunk 1 above)
Comment (rasikasri) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseService.java:101]: Done in RWA-2212 pr

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java [hunk 2]
```diff
@@ -233,17 +391,167 @@ void given_task_is_created_when_count_by_role_signature_then_correct_count_retur
...
+            .cftTaskStates(List.of(CFTTaskState.ASSIGNED, CFTTaskState.COMPLETED))
+            .caseIds(List.of("1623278362430412", "1623278362430413"))
+            .taskTypes(List.of("startAppeal", "reviewAppeal"))
+            .users(List.of("someAssignee", "anotherAssignee"))
+            .build();
+
+        List<String> taskIds = taskResourceRepository.searchTasksIds(0,25, filterSignature, roleSignature,
+            List.of("1623278362430413", "88888888888888"), request);
+        assertEquals(1, taskIds.size());
+        assertEquals(taskId, taskIds.get(0));
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:519]: assert tasks do not have excluded id?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java [hunk 3]
```diff
@@ -233,17 +391,167 @@ void given_task_is_created_when_count_by_role_signature_then_correct_count_retur
...
+            .taskTypes(List.of("startAppeal", "reviewAppeal"))
+            .users(List.of("someAssignee", "anotherAssignee"))
+            .sortingParameters(List.of(new SortingParameter(SortField.CASE_ID, SortOrder.ASCENDANT)))
+            .build();
+
+        List<String> taskIds = taskResourceRepository.searchTasksIds(0,25, filterSignature, roleSignature,
+            List.of(), request);
+        assertEquals(2, taskIds.size());
+        assertEquals(taskId, taskIds.get(0));
+        assertEquals(taskId2, taskIds.get(1));
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:491]: assert tasks contain assigned & unassigned states?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java [hunk 4]
```diff
@@ -233,17 +391,167 @@ void given_task_is_created_when_count_by_role_signature_then_correct_count_retur
...
+            .requestContext(RequestContext.AVAILABLE_TASKS)
+            .caseIds(List.of("1623278362430412", "1623278362430413"))
+            .taskTypes(List.of("startAppeal", "reviewAppeal"))
+            .users(List.of("someAssignee", "anotherAssignee"))
+            .build();
+
+        List<String> taskIds = taskResourceRepository.searchTasksIds(0,25, filterSignature, roleSignature,
+            List.of(), request);
+        assertEquals(1, taskIds.size());
+        assertEquals(taskId2, taskIds.get(0));
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:462]: assert assignee is null?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java [hunk 5]
```diff
@@ -207,23 +215,173 @@ void given_task_is_created_when_find_by_id_then_task_roles_and_work_types_have_e
...
+            .cftTaskStates(List.of(CFTTaskState.ASSIGNED))
+            .caseIds(List.of("1623278362430412", "1623278362430413"))
+            .taskTypes(List.of("startAppeal", "reviewAppeal"))
+            .users(List.of("someAssignee", "anotherAssignee"))
+            .build();
+
+        List<String> taskIds = taskResourceRepository.searchTasksIds(0,25, filterSignature, roleSignature,
+            List.of(), request);
+        assertEquals(1, taskIds.size());
+        assertEquals(taskId, taskIds.get(0));
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:377]: assert state?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java [hunk 6]
```diff
@@ -207,23 +215,173 @@ void given_task_is_created_when_find_by_id_then_task_roles_and_work_types_have_e
...
+            .cftTaskStates(List.of(CFTTaskState.ASSIGNED, CFTTaskState.COMPLETED))
+            .caseIds(List.of("1623278362430412", "1623278362430413"))
+            .taskTypes(List.of("startAppeal", "reviewAppeal"))
+            .users(List.of("someAssignee"))
+            .build();
+
+        List<String> taskIds = taskResourceRepository.searchTasksIds(0,25, filterSignature, roleSignature,
+            List.of(), request);
+        assertEquals(1, taskIds.size());
+        assertEquals(taskId, taskIds.get(0));
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:349]: assert assignee?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java [hunk 7]
```diff
@@ -207,23 +215,173 @@ void given_task_is_created_when_find_by_id_then_task_roles_and_work_types_have_e
...
+            .cftTaskStates(List.of(CFTTaskState.ASSIGNED, CFTTaskState.COMPLETED))
+            .caseIds(List.of("1623278362430412", "1623278362430413"))
+            .taskTypes(List.of("startAppeal"))
+            .users(List.of("someAssignee", "anotherAssignee"))
+            .build();
+
+        List<String> taskIds = taskResourceRepository.searchTasksIds(0,25, filterSignature, roleSignature,
+            List.of(), request);
+        assertEquals(1, taskIds.size());
+        assertEquals(taskId, taskIds.get(0));
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:321]: assert taskType?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java [hunk 8]
```diff
@@ -207,23 +215,173 @@ void given_task_is_created_when_find_by_id_then_task_roles_and_work_types_have_e
...
+            .cftTaskStates(List.of(CFTTaskState.ASSIGNED, CFTTaskState.COMPLETED))
+            .caseIds(List.of("1623278362430412"))
+            .taskTypes(List.of("startAppeal", "reviewAppeal"))
+            .users(List.of("someAssignee", "anotherAssignee"))
+            .build();
+
+        List<String> taskIds = taskResourceRepository.searchTasksIds(0,25, filterSignature, roleSignature,
+            List.of(), request);
+        assertEquals(1, taskIds.size());
+        assertEquals(taskId, taskIds.get(0));
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:293]: assert caseId?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseServiceSearchTest.java [hunk 9]
```diff
@@ -449,6 +427,118 @@ void should_return_ordered_task_list_and_count_when_search_for_all_work() {
...
+
+        GetTasksResponse<Task> response = cftTaskDatabaseService.searchForTasks(0, 25, searchRequest,
+            accessControlResponse, false);
+        assertEquals(2, response.getTotalRecords());
+        Assertions.assertThat(response.getTasks())
+            .hasSize(2)
+            .flatExtracting(Task::getId, Task::getCaseId)
+            .containsExactly(
+                newArrayList(
+                    "8d6cc5cf-c973-11eb-aaaa-000000000001", "1623278362400001",
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseServiceSearchTest.java:505]: asset taskTypes?

Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseServiceSearchTest.java:505]: Done in RWA-2212

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java (hunk 8 above)
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:293]: Done in RWA-2212 in CFTTaskDatabaseServiceSearchTest

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java (hunk 7 above)
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:321]: Done in RWA-2212 in CFTTaskDatabaseServiceSearchTest

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java (hunk 6 above)
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:349]: Done in RWA-2212 in CFTTaskDatabaseServiceSearchTest

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java (hunk 5 above)
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:377]: Done in RWA-2212 in CFTTaskDatabaseServiceSearchTest

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java (hunk 4 above)
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:462]: Done in RWA-2212 in CFTTaskDatabaseServiceSearchTest

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java (hunk 3 above)
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:491]: Done in RWA-2212 in CFTTaskDatabaseServiceSearchTest

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java (hunk 2 above)
Comment (rasikasri) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/repository/TaskResourceRepositoryTest.java:519]: Done in RWA-2212 in CFTTaskDatabaseServiceSearchTest

**Review by serdaroguzhanakin — APPROVED** [review]

**Pull request merged**